File: src/main/scala/ao/learn/mst/gen5/example/RandomPlayer.scala
package ao.learn.mst.gen5.example

import ao.learn.mst.gen5.ExtensivePlayer
import scala.util.Random



class RandomPlayer[InformationSet, Action](
  sourceOfRandomness : Random)
  extends ExtensivePlayer[InformationSet, Action]
{
  def act(informationSet: InformationSet, actions: Traversable[Action]): Action =
  {
    val randomlyWeightedActions : Traversable[(Action, Double)] =
      actions.map((_, sourceOfRandomness.nextDouble()))

    val maximumWeightAction =
      randomlyWeightedActions.maxBy(_._2)._1

    maximumWeightAction
  }
}

File: ai-cfr-java/src/main/java/ao/ai/cfr/sample/RandomPlayer.java
package ao.ai.cfr.sample;

import ao.ai.cfr.ExtensivePlayer;
import ao.ai.cfr.node.choice.ChoiceSet;

import java.util.Random;

public class RandomPlayer<I, A> implements ExtensivePlayer<I, A> {
    private final Random random;

    public RandomPlayer(Random random) {
        this.random = random;
    }


    @Override
    public A act(I i, ChoiceSet<A> as) {
        A maxChoice = null;
        double maxChoiceWeight = 0;

        for (A choice : as) {
            double weight = random.nextDouble();
            if (maxChoiceWeight < weight) {
                maxChoice = choice;
                maxChoiceWeight = weight;
            }
        }

        return maxChoice;
    }
}
